oracle/src/main.sw

```diff
@@ -0,0 +1,58 @@
+contract;
+
+dep interface;
+dep data_structures;
+dep errors;
+dep events;
+
+use std::{
+    chain::auth::msg_sender,
+    constants::BASE_ASSET_ID,
+    identity::Identity,
+    logging::log,
+    result::Result,
+    revert::require,
+};
+
+use interface::Oracle;
+use data_structures::State;
+use errors::{AccessError, InitializationError};
+use events::PriceUpdateEvent;
+
+storage {
+    /// The Identity that can control the oracle (node)
+    owner: Identity = Identity::ContractId(BASE_ASSET_ID),
+    /// Current price of tracked asset
+    price: u64 = 0,
+    /// The initialization state of the contract.
+    state: State = State::NotInitialized,
+}
+
+impl Oracle for Contract {
+    #[storage(read, write)] fn constructor(owner: Identity) {
+        require(storage.state == State::NotInitialized, InitializationError::CannotReinitialize);
+
+        storage.owner = owner;
+        storage.state = State::Initialized;
+    }
+
+    #[storage(read, write)] fn set_price(new_price: u64) {
+        require(storage.state == State::Initialized, InitializationError::ContractNotInitialized);
+        let sender = msg_sender().unwrap();
+        require(sender == storage.owner, AccessError::NotOwner);
+
+        storage.price = new_price;
+
+        log(PriceUpdateEvent {
+            price: new_price
+        });
+    }
+
+    #[storage(read)] fn owner() -> Identity {
+        storage.owner
+    }
+
+    #[storage(read)] fn price() -> u64 {
+        storage.price
```

Comment:
I'm not sure if this is done in Solidity or not, but should you be able to query the price if the oracle hasn't been initialized yet?

Reply:
I was debating whether or not to require the oracle to be initialized.  I decided not to just because 1.) It's unlikely any consumer application will interact with the oracle before it is initialized.  2.) IMO it should be the responsibility of a consumer contract to ensure they are interacting with an initialized contract

Reply:
It may be hard to compare to solidity because solidity constructors are run during contract deployment and not afterwards

Reply:
perhaps `price()` could return an `Option<u64>` ? I don't know if that's an unacceptable gas overhead for the edge case of non-initialized price. But maybe could also be useful if the node encounters a error- then you'd get a `None` instead of a stale price.

Reply:
Options are not yet supported in the SDK so I think the best course of action is to just use a u64 and then change it to and Option in the future.